hbase-client/src/main/java/org/apache/hadoop/hbase/util/PoolMap.java

```diff
@@ -45,177 +36,138 @@
-    return pool != null ? pool.get() : null;
+  public V getOrCreate(K key, PoolResourceSupplier<V> supplier) throws IOException {
+     synchronized (pools) {
+       Pool<V> pool = pools.get(key);
+
+       if (pool == null) {
+         pool = createPool();
+         pools.put(key, pool);
+       }
+
+       try {
+         return pool.getOrCreate(supplier);
+       } catch (IOException | RuntimeException | Error e) {
+         if (pool.size() == 0) {
+           pools.remove(key);
+         }
+
+         throw e;
+       }
+     }
   }
+  public boolean remove(K key, V value) {
+    synchronized (pools) {
+      Pool<V> pool = pools.get(key);
 
-  @Override
-  public V put(K key, V value) {
-    Pool<V> pool = pools.get(key);
-    if (pool == null) {
-      pools.put(key, pool = createPool());
-    }
-    return pool != null ? pool.put(value) : null;
-  }
+      if (pool == null) {
+        return false;
+      }
 
-  @SuppressWarnings("unchecked")
-  @Override
-  public V remove(Object key) {
-    Pool<V> pool = pools.remove(key);
-    if (pool != null) {
-      removeValue((K) key, pool.get());
-    }
-    return null;
-  }
+      boolean removed = pool.remove(value);
 
-  public boolean removeValue(K key, V value) {
-    Pool<V> pool = pools.get(key);
-    boolean res = false;
-    if (pool != null) {
-      res = pool.remove(value);
-      if (res && pool.size() == 0) {
+      if (removed && pool.size() == 0) {
         pools.remove(key);
       }
+
+      return removed;
     }
-    return res;
   }
 
-  @Override
-  public Collection<V> values() {
-    Collection<V> values = new ArrayList<>();
-    for (Pool<V> pool : pools.values()) {
-      Collection<V> poolValues = pool.values();
-      if (poolValues != null) {
-        values.addAll(poolValues);
-      }
+  public void remove(K key) {
+    synchronized (pools) {
+      pools.remove(key);
     }
-    return values;
   }
 
-  public Collection<V> values(K key) {
-    Collection<V> values = new ArrayList<>();
-    Pool<V> pool = pools.get(key);
-    if (pool != null) {
-      Collection<V> poolValues = pool.values();
-      if (poolValues != null) {
-        values.addAll(poolValues);
+  public List<V> values() {
+    List<V> values = new ArrayList<>();
+
+    synchronized (pools) {
+      for (Pool<V> pool : pools.values()) {
+        Collection<V> poolValues = pool.values();
+        if (poolValues != null) {
+          values.addAll(poolValues);
+        }
       }
     }
+
     return values;
   }
 
+  public List<V> values(K key) {
+    synchronized (pools) {
+      Pool<V> pool = pools.get(key);
 
-  @Override
-  public boolean isEmpty() {
-    return pools.isEmpty();
+      if (pool == null) {
+        return Collections.emptyList();
+      } else {
+        return new ArrayList<>(pool.values());
+      }
+    }
   }
 
-  @Override
   public int size() {
-    return pools.size();
+    synchronized (pools) {
+      return pools.size();
+    }
   }
```

Comment:
we still need this? don't see it getting used in `AbstractRpcClient`

Reply:
It was used by `BlockingRpcClient` but only in a trace log, so I removed it.